### PR TITLE
Add service update VCB exchange rates

### PR DESCRIPTION
**Pull request by binhlam**
Add service update VCB exchange rates

### CODE REVIEW

**Review by elicoidal — APPROVED** [review]
LGTM

**Comment by elicoidal** [comment]
Please check travis

**Comment by jcdrubay** [comment]
@elicoidal Actually the issue with Travis was not related to the PR. All good now after I restarted the different builds manually.

**Review by jcdrubay — APPROVED** [review]

**Comment by tungtt** [comment]
Hi,

One tiny issue to fix (according to Travis) and everything should be good.

```
************* Module currency_rate_update.services.update_service_VCB
currency_rate_update/services/update_service_VCB.py:79: [E1601(print-statement), VCBGetter.get_updated_currency] print statement used
```

Thanks for the contribution.

File: currency_rate_update/services/update_service_VCB.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            main_curr_data = self.rate_retrieve(dom, vcb_ns, main_currency)
+
+        for curr in currency_array:
+            self.validate_cur(curr)
+            if curr == 'VND':
+                rate = main_curr_data['rate_currency']
+            else:
+                curr_data = self.rate_retrieve(dom, vcb_ns, curr)
+                if main_currency == 'VND':
+                    print ' === [2]'
```
Comment (hongquangminh) [inline currency_rate_update/services/update_service_VCB.py]: Don't use print on your code :)

Comment (hongquangminh) [reply currency_rate_update/services/update_service_VCB.py]: Opps, The same comment with @tungtt, Sorry!

**Review by jcdrubay — APPROVED** [review]

**Comment by binhlam** [comment]
I have fixed unexpected indentation.

**Review by kenluckiest — APPROVED** [review]
Looks good to me

File: currency_rate_update/services/update_service_VN_VCB.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+# © 2017 Binh Lam
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+"""Add service update VCB exchange rates."""
```
Comment (MiquelRForgeFlow) [inline currency_rate_update/services/update_service_VN_VCB.py]: Shouldn't this be just a comment using `# Add...`?

**Review by MiquelRForgeFlow — APPROVED** [review]
LGTM overall :+1: 
just minor comment

File: currency_rate_update/services/update_service_VN_VCB.py (hunk 2 above)
Comment (binhlam) [reply currency_rate_update/services/update_service_VN_VCB.py]: It's a doc string.

**Review by congdpt — APPROVED** [review]

File: currency_rate_update/services/update_service_VN_VCB.py (hunk 2 above)
Comment (jcdrubay) [reply currency_rate_update/services/update_service_VN_VCB.py]: I think this is duplicated with line 18.

I think a docstring as the first line of the class is more appropriate so you should keep line 18 and remove this line.

**Review by jcdrubay — CHANGES_REQUESTED** [review]

**Comment by MiquelRForgeFlow** [comment]
@jcdrubay please update review

**Review by jcdrubay — APPROVED** [review]

**Comment by jcdrubay** [comment]
@OCA/core-maintainers could this be merged? I think that all should be fine for a merge since September 14th. Thanks.

**Comment by StefanRijnhart** [comment]
How is the CLA status?

**Comment by dreispt** [comment]
I checked and we do have an ICLA for Binh Lam Thai.

**Comment by StefanRijnhart** [comment]
Thanks @dreispt!

**Comment by StefanRijnhart** [comment]
@binhlam great work! It should be ready for merge, but can you rebase your commits after @jcdrubay's one so that we don't have too many fixups in the git log?

**Comment by binhlam** [comment]
@StefanRijnhart  I rebased them. Please check.

**Comment by StefanRijnhart** [comment]
Thanks for the fast response! Tests are broken because the yahoo service was discontinued. For the sake of this PR maybe you can just remove the Yahoo test so that we can have this merged? The Yahoo service is already being removed in the migration to 11.0 https://github.com/OCA/account-financial-tools/pull/579

**Comment by StefanRijnhart** [comment]
Thanks! I think flake8 wants you to remove the blank line left at the end.

**Comment by binhlam** [comment]
@StefanRijnhart  I think it's ok now.

**Comment by StefanRijnhart** [comment]
It's green! Thank you very much!

**Comment by StefanRijnhart** [comment]
This is odd... Tests of the 10.0 main branch with this merged in now fail on the VCB test

```
FAIL: test_cron_VN_VCB (odoo.addons.currency_rate_update.tests.test_currency_rate_update.TestCurrencyRateUpdate)
` Test the ir.cron with Vietcombank service for USD
Traceback (most recent call last):
`   File "/home/travis/build/OCA/account-financial-tools/currency_rate_update/tests/test_currency_rate_update.py", line 81, in test_cron_VN_VCB
`     self._test_cron_by_service('VN_VCB', ['base.USD'])
`   File "/home/travis/build/OCA/account-financial-tools/currency_rate_update/tests/test_currency_rate_update.py", line 44, in _test_cron_by_service
`     self.assertFalse('ERROR' in service_x.note)
` AssertionError: True is not false
FAILED
```

https://travis-ci.org/OCA/account-financial-tools/jobs/306263703#L1444

Any ideas?

**Comment by jcdrubay** [comment]
After re-running the job it passes: https://travis-ci.org/OCA/account-financial-tools/jobs/306263703

I imagine that due to a network issue, the service returned nothing, so this line returned an exception which has been handled as a failure :/

`self.assertFalse('ERROR' in service_x.note)`

Do you have any recommendation on how to handle "unreliable" services?

Maybe we should integrate a retry when the result is None and we could also considered the test as Pass with a logging.warning in case there is no return:

```
if service_x.note:
    self.assertFalse('ERROR' in service_x.note)`
else:
    logging.warning('No value returned from the service.'
```

**Comment by StefanRijnhart** [comment]
@jcdrubay sounds reasonable! But any warning message will still make runbot (not travis) fail. Better make that an info level message.

**Comment by jcdrubay** [comment]
I am really surprised to hear that Runbot would fail but I trust your experience :)

Then it could be something like:
```
if service_x.note:
    self.assertFalse('ERROR' in service_x.note)`
else:
    # Not logging a warning to avoid runbot to fail.
    logging.info('WARNING: No value returned from the service.')  
```

And retry 3 times with 5 seconds interval.

**Comment by StefanRijnhart** [comment]
@jcdrubay You can take my word indeed, or revisit the discussion: https://github.com/OCA/runbot-addons/issues/34

I like your proposal. Maybe not include WARNING in uppercase as this may be how runbot detects the log level (not sure). Will you make a PR?

**Comment by jcdrubay** [comment]
Thanks for your advices. PR will come on Monday from @binhlam Have a nice weekend! :sunny:

**Comment by binhlam** [comment]
@StefanRijnhart Here is my new PR: https://github.com/OCA/account-financial-tools/pull/584

**Pull request merged**